db/db_impl_write.cc

```diff
@@ -1103,42 +1109,67 @@ Status DBImpl::HandleWriteBufferFull(WriteContext* write_context) {
       write_buffer_manager_->buffer_size());
   // no need to refcount because drop is happening in write thread, so can't
   // happen while we're in the write thread
-  ColumnFamilyData* cfd_picked = nullptr;
-  SequenceNumber seq_num_for_cf_picked = kMaxSequenceNumber;
-
-  for (auto cfd : *versions_->GetColumnFamilySet()) {
-    if (cfd->IsDropped()) {
-      continue;
-    }
-    if (!cfd->mem()->IsEmpty()) {
-      // We only consider active mem table, hoping immutable memtable is
-      // already in the process of flushing.
-      uint64_t seq = cfd->mem()->GetCreationSeq();
-      if (cfd_picked == nullptr || seq < seq_num_for_cf_picked) {
-        cfd_picked = cfd;
-        seq_num_for_cf_picked = seq;
+  std::vector<std::vector<uint32_t>> to_flush;
+  // Elements in cfds must be unique.
+  autovector<ColumnFamilyData*> cfds;
+  auto* ifm = static_cast<InternalFlushManager*>(
+      immutable_db_options_.flush_manager.get());
+  assert(ifm != nullptr);
+  if (nullptr == ifm->external_manager_ ||
+      nullptr == ifm->external_manager_->OnHandleWriteBufferFull1) {
+    ifm->OnHandleWriteBufferFull(*versions_->GetColumnFamilySet(), &cfds);
+  } else {
+    ifm->external_manager_->OnHandleWriteBufferFull1(&to_flush);
+    for (const auto& ids : to_flush) {
+      for (const auto id : ids) {
+        bool found = false;
+        ColumnFamilyData* tmp_cfd =
+            versions_->GetColumnFamilySet()->GetColumnFamily(id);
+        for (auto cfd : cfds) {
+          if (cfd == tmp_cfd) {
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          cfds.push_back(tmp_cfd);
+        }
```

Comment:
We can move this inside. ifm->OnHandleWriteBufferFull so that external_manager forwarding is not seen here.

Reply:
According to offline discussion, will try to combine them.